lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,279 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const objectGraph = require('object-graph-js').ObjectGraph;
+
+var noise = false;     // True if base property like "caller", "length" will be returned.
+var showConst = false; // True if constant number/string/boolean value will be returned.
+
+var blacklistProperties = [
+  'toString',    // Do not visit object's toString.
+  'constructor',  // Do not visit constructor functions, they have prototype but not interface.
+  'CSS2Properties' // These are known bugs in Firefox.
+];
+
+var blacklistFunctionAttr = ['prototype'];
+// blacklistFunctionAttr is list of property of Function.prototype,
+// prototype is not a part of function, but we also wish to blacklist it.
+
+var blacklistObjectAttr = [];
+// blacklistObjectAttr is list of property of Object.prototype.
+
+var constType = ["boolean", "number", "string"];
+// Primitive types which is constant will be blacklisted.
+
+var rootId, ObjectId, FunctionId, objectId, functionId;
+
+// A wrapper function of og.getObjectKeys, constant will
+// be filtered if showConst is not turned on.
+function getObjectProperties(og, id) {
+  var keys = og.getObjectKeys(id);
+  keys = keys.filter(key => {
+    var keyId = og.lookup(key, id);
+    var meta = og.lookupMetaData(key, id);
+    if (Object.keys(meta).length === 0) {
+      // If meta contains no information, it means this property
+      // is inherited from its prototype. Inherited properties
+      // will be added to thier prototype interface.
+      return false;
+    }
+    if (!showConst) {
+      // If showConst is not turned on, filter the result if
+      // string it is premetive type and writable is false.
+      return constType.indexOf(og.getType(keyId)) === -1 || meta.writable !== 0;
+    }
+    return true;
+  });
+  return keys;
+}
+
+// Remove all +, $ chars in given string, return the new string.
+function cleanUp(str) {
+  return str.replace(/\+/g, '').replace(/\$/g, '');
+}
+
+// Return if str is a positive interger.
+function isPosotiveInt(str) {
+  return /^\+?(0|[1-9]\d*)$/.test(str);
+}
+
+// Merge all given arrays into arr, no dupulicate allowed,
+// blacklisted properties are filtered.
+function arrayMerge(arr, opt) {
+  opt = opt || {};
+
+  for (var i = 1; i < arguments.length; i += 1) {
+    for (var j = 0; j < arguments[i].length; j += 1) {
+      var str = cleanUp(arguments[i][j]);
+      if (arr.indexOf(str) === -1 &&
+        ((blacklistObjectAttr.indexOf(str) === -1 &&
+        blacklistFunctionAttr.indexOf(str) === -1) || noise) &&
+        !isPosotiveInt(str)) {
```

Comment:
Is `isPositiveInt` a proxy for "this looks like it's an object graph ID that has been turned into a string"? If so (and we don't use it in other ways), we should give `isPositiveInt` a name that describes how it's used instead of the details of how it works.